### PR TITLE
Added ability to create db with NopWal

### LINKED ISSUE #452: Ability to create db with NopWAL
We create a test db in prometheus for unit-testing and general tests. Ideally we never write anything to disk there but turns out we do because we cannot disable the WAL.

### PR DESCRIPTION
**Pull request by vn-ki**
Closes #452 

Signed-off-by: Vishnunarayan K I <appukuttancr@gmail.com>

### CODE REVIEW

**Comment by krasi-georgiev** [comment]
maybe can use the segmentSize ? when set to 0 it would be essentially No Wal.

**Comment by vn-ki** [comment]
> maybe can use the segmentSize ? when set to 0 it would be essentially No Wal.

@krasi-georgiev When I looked the function `wal.NewSize`, I saw that it created directories. The current implementation avoids that. I'm not sure what would happen if `segmentSize` is set to 0. What do you think? Should I change it to use `segmentSize`?

**Comment by vn-ki** [comment]
@krasi-georgiev Updated. But the test errors still exists.

**Comment by krasi-georgiev** [comment]
leave the test for the end, but the logic is wrong.

we should have 3 cases - default wal size , no wal, custom wal size.

**Comment by vn-ki** [comment]
> we should have 3 cases - default wal size , no wal, custom wal size.

I'm sorry. I don't understand. I have to add `NopWal` back in? And chack that as well as `WalSegmentSize`? 

What is the third case?

**Comment by krasi-georgiev** [comment]
I think something like this will work 
db.opts.WALSegmentSize == 0 : use the default size
db.opts.WALSegmentSize == -1 : wal is nil or disabled
db.opts.WALSegmentSize > 0 : wal is set to custom size.


about the test dig deeper and if you still can't figure it out by the end of next week ping me and I will try to help.

**Comment by vn-ki** [comment]
I have updated it. Please check if this is what you mean.

> about the test dig deeper and if you still can't figure it out by the end of next week ping me and I will try to help.

Okay. Thank you.

I took a look and I think it is because some tests check for the wal files that is created. Two of the failed tests create tmp dirs for this check.

Should I remove these checks (if they're tested in `wal_test.go`) or make a different `db.Option` with wal for these tests?

**Comment by krasi-georgiev** [comment]
> I have updated it. Please check if this is what you mean.

yes and also update the comment in the options about the 3 possible cases.
https://github.com/prometheus/tsdb/pull/519/files#diff-7a117cbfd5485ed35efcbb2995c37aa0R55

>I took a look and I think it is because some tests check for the wal files that is created. Two of the failed tests create tmp dirs for this check.
Should I remove these checks (if they're tested in `wal_test.go`) or make a different `db.Option` with wal for these tests?

Without looking at the tests can't tell, but study the code and if you are still unsure next week ping me and I will try to help.

**Comment by krasi-georgiev** [comment]
also don't forget that need to add a test to ensure this behaviour - setting the no WAL option should not create a WAL file.

maybe can add it to 
https://github.com/prometheus/tsdb/blob/9f28ffa6f40a3d79a3eca28c775324eb65d354a4/db_test.go#L749

**Comment by krasi-georgiev** [comment]
ping me when this is ready for another review

**Comment by vn-ki** [comment]
@krasi-georgiev Please take a look.

**Comment by vn-ki** [comment]
@krasi-georgiev The travis seems unrelated.

```
Attempt to get headers for https://sourceforge.net/projects/ezwinports/files/make-4.2.1-without-guile-w32-bin.zip/download failed.
  The remote file either doesn't exist, is unauthorized, or is forbidden for url 'https://sourceforge.net/projects/ezwinports/files/make-4.2.1-without-guile-w32-bin.zip/download'. Exception calling "GetResponse" with "0" argument(s): "Unable to connect to the remote server"
Downloading make 64 bit
  from 'https://sourceforge.net/projects/ezwinports/files/make-4.2.1-without-guile-w32-bin.zip/download'
ERROR: The remote file either doesn't exist, is unauthorized, or is forbidden for url 'https://sourceforge.net/projects/ezwinports/files/make-4.2.1-without-guile-w32-bin.zip/download'. Exception calling "GetResponse" with "0" argument(s): "Unable to connect to the remote server" 
This package is likely not broken for licensed users - see https://chocolatey.org/docs/features-private-cdn.
  make may be able to be automatically uninstalled.
The install of make was NOT successful.
Error while running 'C:\ProgramData\chocolatey\lib\make\tools\chocolateyinstall.ps1'.

```

**Comment by krasi-georgiev** [comment]
I would say better to refactor TestWALSegmentSizeOption to check all 3  WALSegmentSize variants.

The required setup can be wrapped in a bootstrap function which takes the wal options as a part and check that the output matched the expected wal segment size.

have a look at `TestDeleteCompactionBlockAfterFailedReload`

**Comment by krasi-georgiev** [comment]
@vn-ki are you still willing to continue the work on this PR?

**Comment by vn-ki** [comment]
@krasi-georgiev Yes, I'm still interested in working on this. I'm having mid sems now. So got caught up in that. I'll look into this later today.

File: db.go [hunk 1]
```diff
@@ -267,9 +270,14 @@ func Open(dir string, l log.Logger, r prometheus.Registerer, opts *Options) (db
 	if opts.WALSegmentSize > 0 {
 		segmentSize = opts.WALSegmentSize
 	}
-	wlog, err := wal.NewSize(l, r, filepath.Join(dir, "wal"), segmentSize)
-	if err != nil {
-		return nil, err
+	var wlog *wal.WAL
+	if opts.WALSegmentSize == -1 {
+		wlog = nil
```
Comment (krasi-georgiev) [inline db.go]: can drop this as `var wlog *wal.WAL` sets it to nil anyway.

can be just 
```
// check that the wal is not disabled.
if opts.WALSegmentSize != -1 {
	wlog, err = wal.NewSize(l, r, filepath.Join(dir, "wal"), segmentSize)
	if err != nil {
		return nil, err
	}
}
```

**Comment by krasi-georgiev** [comment]
@vn-ki 	when do you think you can continue with this as it is blocking PR https://github.com/prometheus/tsdb/pull/541 ?

**Comment by vn-ki** [comment]
I don't think I can look into this before this weekend. Sorry for blocking
the other PR. Feel free to close this if you want the other PR merged
before the weekend.

**Comment by krasi-georgiev** [comment]
The weekend is fine I think.

**Comment by vn-ki** [comment]
@krasi-georgiev I have updated the PR. Please see if this is what you meant.

File: db_test.go [hunk 2]
```diff
@@ -743,28 +743,44 @@ func TestWALFlushedOnDBClose(t *testing.T) {
...
+			for i, f := range files {
+				if len(files)-1 != i {
+					testutil.Equals(t, int64(options.WALSegmentSize), f.Size(), "WAL file size doesn't match WALSegmentSize option, filename: %v", f.Name())
+					continue
+				}
+				testutil.Assert(t, int64(options.WALSegmentSize) > f.Size(), "last WAL file size is not smaller than the WALSegmentSize option, filename: %v", f.Name())
+			}
+		},
+		-1: func(dbDir string, options Options) {
+			if _, err := os.Stat(filepath.Join(dbDir, "wal")); os.IsNotExist(err) {
```
Comment (krasi-georgiev) [inline db_test.go]: can't this be something like: 
```
if _, err := os.Stat(filepath.Join(dbDir, "wal")); !os.IsNotExist(err) {
	t.Fatal("wal directory is present when the wal is disabled")
}
```

File: db_test.go [hunk 3]
```diff
@@ -743,28 +743,44 @@ func TestWALFlushedOnDBClose(t *testing.T) {
...
-	options.WALSegmentSize = 2 * 32 * 1024
-	db, delete := openTestDB(t, &options)
-	defer delete()
-	app := db.Appender()
-	for i := int64(0); i < 155; i++ {
-		_, err := app.Add(labels.Labels{labels.Label{Name: "wal", Value: "size"}}, i, rand.Float64())
-		testutil.Ok(t, err)
-		testutil.Ok(t, app.Commit())
+	tests := map[int]func(dbdir string, options Options){
+		2 * 32 * 1024: func(dbDir string, options Options) {
```
Comment (krasi-georgiev) [inline db_test.go]: some comments for each test case would be a nice reminder.

// Custom wal size.

// Default wal size.

// WAL is disabled.

**Comment by vn-ki** [comment]
@krasi-georgiev Done

File: db_test.go [hunk 4]
```diff
@@ -743,28 +743,56 @@ func TestWALFlushedOnDBClose(t *testing.T) {
...
+		// Test Default Wal Size
+		0: func(dbDir string, options Options) {
+			files, err := ioutil.ReadDir(filepath.Join(dbDir, "wal"))
+			testutil.Ok(t, err)
+			for i, f := range files {
+				if len(files)-1 != i {
+					testutil.Equals(t, int64(DefaultOptions.WALSegmentSize), f.Size(), "WAL file size doesn't match WALSegmentSize option, filename: %v", f.Name())
+					continue
+				}
+				testutil.Assert(t, int64(DefaultOptions.WALSegmentSize) > f.Size(), "last WAL file size is not smaller than the WALSegmentSize option, filename: %v", f.Name())
```
Comment (krasi-georgiev) [inline db_test.go]: can be simplified a bit for both cases.

```
for _, f := range files[:len(files)-1] {
	testutil.Equals(t, int64(DefaultOptions.WALSegmentSize), f.Size(), "WAL file size doesn't match WALSegmentSize option, filename: %v", f.Name())
}
lastFile := files[len(files)-1]
testutil.Assert(t, int64(DefaultOptions.WALSegmentSize) > lastFile.Size(), "last WAL file size is not smaller than the WALSegmentSize option, filename: %v", lastFile.Name())
```

File: db.go [hunk 5]
```diff
@@ -292,9 +295,13 @@ func Open(dir string, l log.Logger, r prometheus.Registerer, opts *Options) (db
 	if opts.WALSegmentSize > 0 {
 		segmentSize = opts.WALSegmentSize
 	}
-	wlog, err := wal.NewSize(l, r, filepath.Join(dir, "wal"), segmentSize)
-	if err != nil {
-		return nil, err
+	fmt.Printf("%d", segmentSize)
```
Comment (krasi-georgiev) [inline db.go]:
```suggestion
```

File: db_test.go [hunk 6]
```diff
@@ -743,28 +743,56 @@ func TestWALFlushedOnDBClose(t *testing.T) {
...
+			}
+		},
+		// Test Custom Wal Size
```
Comment (krasi-georgiev) [inline db_test.go]: full stops for all comments and no need to add `Test` for all comments.
```suggestion
		// Custom Wal Size.
```

**Comment by vn-ki** [comment]
@krasi-georgiev Done

File: db_test.go [hunk 7]
```diff
@@ -743,28 +743,52 @@ func TestWALFlushedOnDBClose(t *testing.T) {
...
-	options := *DefaultOptions
-	options.WALSegmentSize = 2 * 32 * 1024
-	db, delete := openTestDB(t, &options)
-	defer delete()
-	app := db.Appender()
-	for i := int64(0); i < 155; i++ {
-		_, err := app.Add(labels.Labels{labels.Label{Name: "wal", Value: "size"}}, i, rand.Float64())
-		testutil.Ok(t, err)
-		testutil.Ok(t, app.Commit())
+	tests := map[int]func(dbdir string, options Options){
```
Comment (krasi-georgiev) [inline db_test.go]: can just use the segmentSize as an int, no need to pass all options.

```suggestion
	tests := map[int]func(dbdir string, segmentSize int){
```

File: CHANGELOG.md [hunk 8]
```diff
@@ -1,5 +1,6 @@
 ## master / unreleased
  - [REMOVED] `FromData` is considered unused so was removed.
+ - [FEATURE] Feature to create db with WAL.
```
Comment (krasi-georgiev) [inline CHANGELOG.md]:
```suggestion
 - [FEATURE] Added option WALSegmentSize -1 to disable the WAL.
```

File: db_test.go [hunk 9]
```diff
@@ -743,28 +743,52 @@ func TestWALFlushedOnDBClose(t *testing.T) {
 }
 
 func TestWALSegmentSizeOption(t *testing.T) {
```
Comment (krasi-georgiev) [inline db_test.go]:
```suggestion
func TestWALSegmentSizeOptions(t *testing.T) {
```

File: db.go [hunk 10]
```diff
@@ -55,6 +55,9 @@ var DefaultOptions = &Options{
 // Options of the DB storage.
 type Options struct {
 	// Segments (wal files) max size
+	// When WALSegmentSize is 0, segment size is default size
```
Comment (krasi-georgiev) [inline db.go]:
```
// Segments (wal files) max size.
// WALSegmentSize = 0, segment size is default size.
// WALSegmentSize > 0, segment size is WALSegmentSize.
// WALSegmentSize = -1, wal is disabled.
```
full stops and removed some repetitive words.

**Comment by krasi-georgiev** [comment]
LGTM

ping @codesome

File: db.go [hunk 11]
```diff
@@ -54,7 +54,10 @@ var DefaultOptions = &Options{
 
 // Options of the DB storage.
 type Options struct {
-	// Segments (wal files) max size
+	// Segments (wal files) max size.
+	// WALSegmentSize = 0, segment size is default size.
```
Comment (codesome) [inline db.go]: Right now it takes the default size for `WALSegmentSize < -1` too. Should we make it disabled for `WALSegmentSize < 0` then?

Comment (krasi-georgiev) [reply db.go]: yep , not a bad idea. I like it.

File: db.go [hunk 12]
```diff
@@ -292,9 +295,13 @@ func Open(dir string, l log.Logger, r prometheus.Registerer, opts *Options) (db
 	if opts.WALSegmentSize > 0 {
 		segmentSize = opts.WALSegmentSize
 	}
-	wlog, err := wal.NewSize(l, r, filepath.Join(dir, "wal"), segmentSize)
-	if err != nil {
-		return nil, err
+	var wlog *wal.WAL
+	if opts.WALSegmentSize >= 0 {
```
Comment (krasi-georgiev) [inline db.go]:
```
	// Wal is enabled.
	if opts.WALSegmentSize >= 0 {
		// Wal is set to a custom size.
		if opts.WALSegmentSize > 0 {
			segmentSize = opts.WALSegmentSize
		}	
		wlog, err = wal.NewSize(l, r, filepath.Join(dir, "wal"), segmentSize)
		if err != nil {
			return nil, err
		}	
	}
```

Comment (vn-ki) [reply db.go]: I'm so sorry. I somehow did not see this. I have updated.

**Comment by krasi-georgiev** [comment]
Thanks

**Pull request merged**